pkg/cmd/roachprod-microbench/cluster/execute.go

```diff
@@ -86,52 +132,76 @@ func remoteWorker(
 // parameter indicates whether the execution should stop on the first error.
 func ExecuteRemoteCommands(
 	log *logger.Logger,
+	execFunc RemoteExecutionFunc,
 	cluster string,
 	commandGroups [][]RemoteCommand,
 	numNodes int,
 	failFast bool,
 	callback func(response RemoteResponse),
-) {
+) error {
 	workChannel := make(chan []RemoteCommand, numNodes)
 	responseChannel := make(chan RemoteResponse, numNodes)
 
-	ctx, cancelCtx := context.WithCancel(context.Background())
+	ctx, cancelCtx := context.WithCancelCause(context.Background())
 
 	for idx := 1; idx <= numNodes; idx++ {
-		go remoteWorker(ctx, log, fmt.Sprintf("%s:%d", cluster, idx), workChannel, responseChannel)
+		go remoteWorker(ctx, log, execFunc, fmt.Sprintf("%s:%d", cluster, idx), workChannel, responseChannel)
 	}
 
 	var wg sync.WaitGroup
 	// Receive responses.
 	go func() {
+		nodesAlive := numNodes
 		for {
 			response := <-responseChannel
-			if response.Args == nil {
+			if response.Args == nil && response.commandStatus == 0 { // channel closed
```

Comment:
We could use `Completed` instead of `0` for clarity. I don't get the "channel is closed" comment.